### PR TITLE
Fix `ReadPendingException` warning

### LINKED ISSUE #13177: Avoid to log a warning when unneeded in org.eclipse.jetty.io.FillInterest#register/org.eclipse.jetty.io.FillInterest#tryRegister
**Jetty version(s)**
12.0

**Enhancement Description**
org.eclipse.jetty.io.FillInterest#tryRegister return false if _interested can't be set to `callback` from `null` but there is a case, when you forward the same stream (or Flow.Publisher) in request payload to the output (use case is indirect but ends like that), where you might end up registering the same callback twice (once for the input and once for the output) so it will return false, caller (org.eclipse.jetty.io.FillInterest#register) will log a warning, org.eclipse.jetty.util.thread.SerializedInvoker.Link#run will catch it and call `onError` which just logs yet another warning.

So functionally all works fine since the interests will be the same ones etc but you get warnings which is bothering.

There are a lot of potential fixes but one is to not return false if the already registered callback is the requested callback.

In pseudo code:

```java
        if (!_interested.compareAndSet(null, callback))
            return _interested.get() == callback;
```

but I assume a more correct version is:

```java
        if (_interested.updateAndGet(before -> before == null ? callback : before) != callback)
            return false;
```

### PR DESCRIPTION
**Pull request by lorban**
The problem is that `HttpReceiverOverHTTP.parse()` calls `responseSuccess()` assuming that this method will queue its task into the `HttpReceiver`'s serialized executor for later execution.

Unfortunately, if `Response.Listener.onContentSource()` spawns a thread that does the usual read/demand loop on the given `Content.Source` instance, `HttpReceiverOverHTTP.parse()`'s call to `responseSuccess()` immediately executes the task which calls `HttpReceiver.reset()` before returning. This means that the `Content.Source.read()` call that initiated the `HttpReceiverOverHTTP.parse()` call does not see the `EOF` chunk but null, so it thinks it must demand while `responseSuccess()` already called `receiveNext()` which also did demand.

Fixes #13177

### CODE REVIEW

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java [hunk 1]
```diff
@@ -126,6 +127,7 @@ public Content.Chunk read(boolean fillInterestIfNeeded)
         if (LOG.isDebugEnabled())
             LOG.debug("ParseAndFill needFillInterest {} in {}", needFillInterest, this);
         chunk = consumeChunk();
+        runCompletionTaskIfAny();
```
Comment (sbordet) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java]: This aims to be generic, but it does not need to.
I think it's clearer to call `responseSuccess()` directly from here.

File: jetty-core/jetty-fcgi/jetty-fcgi-client/src/main/java/org/eclipse/jetty/fcgi/client/transport/internal/HttpReceiverOverFCGI.java [hunk 2]
```diff
@@ -88,6 +88,7 @@ public Content.Chunk read(boolean fillInterestIfNeeded)
         HttpConnectionOverFCGI httpConnection = getHttpConnection();
         boolean needFillInterest = httpConnection.parseAndFill(false);
         chunk = consumeChunk();
+        httpConnection.runCompletionTaskIfAny();
```
Comment (sbordet) [inline jetty-core/jetty-fcgi/jetty-fcgi-client/src/main/java/org/eclipse/jetty/fcgi/client/transport/internal/HttpReceiverOverFCGI.java]: Similarly, can we just call `responseSuccess()` from here?

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientTest.java [hunk 3]
```diff
@@ -79,6 +80,158 @@
...
+                    else
+                        body.complete("");
+                }
+                ).start();
+            }
+        };
+
+        client.newRequest(server.getURI())
+            .method("POST")
+            .send(response);
```
Comment (sbordet) [inline jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientTest.java]: We need to parametrize this test with zero content, but also some content.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java (hunk 1 above)
Comment (lorban) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java]: Done.

File: jetty-core/jetty-fcgi/jetty-fcgi-client/src/main/java/org/eclipse/jetty/fcgi/client/transport/internal/HttpReceiverOverFCGI.java (hunk 2 above)
Comment (lorban) [reply jetty-core/jetty-fcgi/jetty-fcgi-client/src/main/java/org/eclipse/jetty/fcgi/client/transport/internal/HttpReceiverOverFCGI.java]: Done.

File: jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientTest.java (hunk 3 above)
Comment (lorban) [reply jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientTest.java]: This is done in another test, I've refactored the newly added ones to make them clearer.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java [hunk 4]
```diff
@@ -359,10 +361,14 @@ private boolean parse(boolean notifyContentAvailable)
...
                                 LOG.debug("Discarding unexpected content after response {}: {} in {}", status, BufferUtil.toDetailString(byteBuffer), this);
                             BufferUtil.clear(byteBuffer);
-                            return false;
                         }
                     }
 
+                    if (notifyContentAvailable)
+                        responseSuccess(exchange, receiveNext);
+                    // else: Let read(boolean) call responseSuccess() so it gets a chance to read this.chunk before
+                    // responseSuccess() resets it to null.
```
Comment (sbordet) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java]:
```suggestion
                    // When notifyContentAvailable==false, this method is called from read(boolean),
                    // and the call to responseSuccess() is performed by read().
                    if (notifyContentAvailable)
                        responseSuccess(exchange, receiveNext);
```

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java (hunk 4 above)
Comment (lorban) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpReceiverOverHTTP.java]: Done.

**Review by sbordet — APPROVED** [review]
